agent/grpc-internal/balancer/custombalancer.go

```diff
@@ -0,0 +1,87 @@
+package balancer
+
+import (
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/resolver"
+)
+
+func init() {
+	balancer.Register(newCustomPickfirstBuilder())
+}
+
+// logger is referenced in pickfirst.go.
+// The gRPC library uses the same component name.
+var logger = grpclog.Component("balancer")
+
+func newCustomPickfirstBuilder() balancer.Builder {
+	return &customPickfirstBuilder{}
+}
+
+type customPickfirstBuilder struct{}
+
+func (*customPickfirstBuilder) Build(cc balancer.ClientConn, opt balancer.BuildOptions) balancer.Balancer {
+	return &customPickfirstBalancer{
+		pickfirstBalancer: pickfirstBalancer{cc: cc},
+	}
+}
+
+func (*customPickfirstBuilder) Name() string {
+	return "pick_first_custom"
+}
+
+// customPickfirstBalancer overrides UpdateClientConnState of pickfirstBalancer.
+type customPickfirstBalancer struct {
+	pickfirstBalancer
+
+	activeAddr resolver.Address
+}
+
+func (b *customPickfirstBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	for _, a := range state.ResolverState.Addresses {
+		// This hack preserves an existing behavior in our client-side
+		// load balancing where if the first address in a shuffled list
+		// of addresses matched the currently connected address, it would
+		// be an effective no-op.
+		if a.Equal(b.activeAddr) {
+			return nil
```

Comment:
TODO: This should be a `break` so that `UpdateClientConnState(state)` is still called with a potentially different list of addresses.

Reply:
https://github.com/hashicorp/consul/pull/15844